File: TesterApiYiqi/descargar_reporte.py
import urllib.request
import urllib.parse
import http.cookiejar
import sys
import os

# Credenciales de YiQi
USERNAME = "mercadolibre@tmcrespo.com.ar"
PASSWORD = "AdministracionMessi"

def descargar_reporte_pdf(schema_id, report_id, instance_id, output_filename):
    """
    Inicia sesión en me.yiqi.com.ar mediante cookies y descarga el PDF de un reporte.
    """
    print(f"Iniciando sesión en me.yiqi.com.ar para {USERNAME}...")
    
    # 1. Configurar el manejador de cookies automático
    cj = http.cookiejar.CookieJar()
    opener = urllib.request.build_opener(urllib.request.HTTPCookieProcessor(cj))
    
    # 2. Realizar POST de Login
    login_url = "https://me.yiqi.com.ar/Account/Login?ReturnUrl=%2F"
    login_data = urllib.parse.urlencode({
        "UserName": USERNAME,
        "Password": PASSWORD,
        "RememberMe": "true",
        "sid": ""
    }).encode("utf-8")
    
    try:
        req = urllib.request.Request(login_url, data=login_data, method="POST")
        req.add_header("Content-Type", "application/x-www-form-urlencoded")
        
        with opener.open(req) as resp:
            # Si el inicio de sesión es exitoso, redirecciona a la raíz
            if resp.status == 200:
                print("¡Inicio de sesión exitoso!")
            else:
                print(f"Advertencia: Código de respuesta de login: {resp.status}")
    except Exception as e:
        print(f"Error al iniciar sesión: {e}")
        return False

    # Mostrar cookies obtenidas
    print("Sesión establecida. Cookies obtenidas:")
    for cookie in cj:
        print(f"  - {cookie.name}: {cookie.value[:15]}...")
        
    # 3. Solicitar el reporte PDF con la sesión activa (cookies)
    report_url = f"https://me.yiqi.com.ar/report/view?schemaId={schema_id}&reportId={report_id}&instanceId={instance_id}"
    print(f"Descargando reporte desde: {report_url}")
    
    try:
        report_req = urllib.request.Request(report_url, method="GET")
        with opener.open(report_req) as resp:
            content_type = resp.headers.get("Content-Type", "")
            if "application/pdf" not in content_type:
                print(f"Error: La respuesta no es un PDF. Content-Type: {content_type}")
                # Imprimir los primeros caracteres por si es un error HTML
                body = resp.read()
                print(body[:200].decode("utf-8", errors="ignore"))
                return False
                
            pdf_data = resp.read()
            print(f"PDF descargado con éxito ({len(pdf_data)} bytes).")
            
            # Guardar el archivo
            with open(output_filename, "wb") as f:
                f.write(pdf_data)
            print(f"Archivo guardado como: {os.path.abspath(output_filename)}")
            return True
            
    except Exception as e:
        print(f"Error al descargar el reporte: {e}")
        return False

if __name__ == "__main__":
    # Parámetros del ejemplo del usuario:
    # Pedido: https://me.yiqi.com.ar/view/PEDIDO?schemaId=1491#/27962
    # Notificación de pedido: reportId=136, instanceId=27962, schemaId=1491
    schema = 1491
    report = 136
    instance = 27962
    output = "notificacion_pedido_27962.pdf"
    
    descargar_reporte_pdf(schema, report, instance, output)

File: TesterApiYiqi/descargar_reporte.js
const fs = require('fs');
const path = require('path');

// Credenciales de YiQi
const USERNAME = "mercadolibre@tmcrespo.com.ar";
const PASSWORD = "AdministracionMessi";

async function descargarReportePdf(schemaId, reportId, instanceId, outputFilename) {
    console.log(`Iniciando sesión en me.yiqi.com.ar para ${USERNAME}...`);
    
    const loginUrl = "https://me.yiqi.com.ar/Account/Login?ReturnUrl=%2F";
    const body = new URLSearchParams({
        UserName: USERNAME,
        Password: PASSWORD,
        RememberMe: "true",
        sid: ""
    });
    
    try {
        // 1. Realizar POST de Login. Usamos redirect: 'manual' para evitar seguir la redirección
        // y poder extraer las cabeceras Set-Cookie directamente.
        const loginResponse = await fetch(loginUrl, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded'
            },
            body: body,
            redirect: 'manual'
        });
        
        // Extraer las cookies devueltas por el servidor
        // En Node.js v18+, fetch soporta headers.getSetCookie()
        let cookies = [];
        if (typeof loginResponse.headers.getSetCookie === 'function') {
            cookies = loginResponse.headers.getSetCookie();
        } else {
            // Fallback para entornos donde no esté getSetCookie
            const rawCookies = loginResponse.headers.get('set-cookie');
            if (rawCookies) {
                cookies = rawCookies.split(',').map(c => c.trim());
            }
        }
        
        if (cookies.length === 0) {
            console.error("No se recibieron cookies del servidor. Verificá las credenciales.");
            return false;
        }
        
        // Armar el string para la cabecera Cookie
        const cookieHeader = cookies.map(c => c.split(';')[0]).join('; ');
        console.log("¡Sesión establecida con éxito!");
        
        // 2. Descargar el reporte PDF usando las cookies obtenidas
        const reportUrl = `https://me.yiqi.com.ar/report/view?schemaId=${schemaId}&reportId=${reportId}&instanceId=${instanceId}`;
        console.log(`Descargando reporte desde: ${reportUrl}`);
        
        const reportResponse = await fetch(reportUrl, {
            method: 'GET',
            headers: {
                'Cookie': cookieHeader
            }
        });
        
        if (!reportResponse.ok) {
            console.error(`Error HTTP al descargar reporte: ${reportResponse.status}`);
            const errText = await reportResponse.text();
            console.error(errText.substring(0, 500));
            return false;
        }
        
        const contentType = reportResponse.headers.get('content-type') || '';
        if (!contentType.includes('application/pdf')) {
            console.error(`Error: El contenido recibido no es un PDF. Content-Type: ${contentType}`);
            const text = await reportResponse.text();
            console.error(text.substring(0, 500));
            return false;
        }
        
        const arrayBuffer = await reportResponse.arrayBuffer();
        const buffer = Buffer.from(arrayBuffer);
        
        fs.writeFileSync(outputFilename, buffer);
        console.log(`¡PDF guardado con éxito! Guardado en: ${path.resolve(outputFilename)}`);
        return true;
        
    } catch (error) {
        console.error("Error en la ejecución:", error);
        return false;
    }
}

// Ejecutar ejemplo
// Pedido: https://me.yiqi.com.ar/view/PEDIDO?schemaId=1491#/27962
// Notificación de pedido: reportId=136, instanceId=27962, schemaId=1491
const schema = 1491;
const report = 136;
const instance = 27962;
const output = "notificacion_pedido_27962_node.pdf";

descargarReportePdf(schema, report, instance, output).catch(console.error);
